src/components/Personalization/utils/surfaceUtils.js

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import {
+  WEB,
+  WEBAPP,
+  SURFACE_TYPE_DELIMITER,
+  FRAGMENT_DELIMITER
+} from "../constants/surface";
+import { startsWith, isNil, isNonEmptyString, includes } from "../../../utils";
+
+const SURFACE_REGEX = /^(\w+):\/\/([^/#]+)(\/[^#]*)?(#.*)?$/;
+const AUTHORITY_REGEX = /^(?:.*@)?(?:[a-z\d\u00a1-\uffff.-]+|\[[a-f\d:]+])(?::\d+)?$/;
+const PATH_REGEX = /^\/(?:[/\w\u00a1-\uffff-.~]|%[a-fA-F\d]{2})*$/;
+const FRAGMENT_REGEX = /^#(?:[/\w\u00a1-\uffff-.~]|%[a-fA-F\d]{2})+$/;
+
+const normalizePath = (path = "/") => {
+  let end = path.length;
+  while (end > 0 && "/".indexOf(path.charAt(end - 1)) !== -1) {
+    end -= 1;
+  }
+  return path.substring(0, end) || "/";
+};
+
+const parseSurface = surfaceString => {
+  const matched = surfaceString.match(SURFACE_REGEX);
```

Comment:
This looks really, really, really bad. Could we use `if/else` and maybe extract some code to helper functions. Otherwise this code is pretty cryptic.

Reply:
Also instead of calling `toLowerCase()` you could probably, make sure the initial value is lower cased.

Reply:
Sure, I've extracted helper functions
Regarding `toLowerCase()` - I can't make the whole initial value lower case, since path and fragment should still remain case-sensitive

Reply:
is it the fragment case-sensitive at authoring mode too?

Reply:
Fragments are not yet supported in the UI, but they are expected to be case-sensitive, yes